### PR TITLE
New Rule: Auth0 MFA Policy Enabled

### PR DESCRIPTION
**Pull request by kleon919**
### Background

[[Hackathon] auth0.mfa.policy.enabled](https://www.notion.so/pantherlabs/Hackathon-auth0-mfa-policy-enabled-b248f09294ec414490bf3f9258ae2aee?pvs=4)

### Changes

This PR adds a medium severity rule to capture when a user enabled an MFA policy.

### Testing

```text
user@computer:~/Sources/PAN/panther-analysis $ pipenv run panther_analysis_tool
test --filter RuleID=Auth0.MFA.Policy.Enabled
[INFO][root]: Testing analysis items in .

Auth0.MFA.Policy.Enabled
	[PASS] MFA Policy Enabled First
		[PASS] [rule] true
		[PASS] [title] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Always Require].
		[PASS] [dedup] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Always Require].
		[PASS] [alertContext] {"actor": {"email": "user.name@yourcompany.io", "name": "User Name", "user_id": "google-oauth2|105261262156475850461"}, "action": "Set the Multi-factor Authentication policies"}
	[PASS] MFA Policy Enabled Second
		[PASS] [rule] true
		[PASS] [title] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Use Adaptive MFA].
		[PASS] [dedup] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Use Adaptive MFA].
		[PASS] [alertContext] {"actor": {"email": "user.name@yourcompany.io", "name": "User Name", "user_id": "google-oauth2|105261262156475850461"}, "action": "Set the Multi-factor Authentication policies"}
	[PASS] Other Event
		[PASS] [rule] false

--------------------------
Panther CLI Test Summary
	Path: .
	Passed: 1
	Failed: 0
	Invalid: 0
```

### CODE REVIEW

File: rules/auth0_rules/auth0_mfa_policy_enabled.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        [
+            data_description == "Set the Multi-factor Authentication policies",
+            request_path == "/api/v2/guardian/policies",
+            is_auth0_config_event(event),
+        ]
+    )
+
+
+def title(event):
+
```
Comment (andrea-youwakim) [inline rules/auth0_rules/auth0_mfa_policy_enabled.py:24]: please also remove this unnecessary space

File: rules/auth0_rules/auth0_mfa_policy_enabled.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    )
+    request_body = deep_get(event, "data", "details", "request", "body", default=[])
+
+    if "all-applications" in request_body:
+        setting_change = "Always Require"
+    if "confidence-score" in request_body:
+        setting_change = "Use Adaptive MFA"
+
+    return (
+        f"Auth0 user {user_email} set the "
```
Comment (andrea-youwakim) [inline rules/auth0_rules/auth0_mfa_policy_enabled.py:36]:
```suggestion
        f"Auth0 user [{user_email}] set the "
```

File: rules/auth0_rules/auth0_mfa_policy_enabled.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        f"mfa policies in your organization to {setting_change}."
```
Comment (andrea-youwakim) [inline rules/auth0_rules/auth0_mfa_policy_enabled.py:37]:
```suggestion
        f"mfa policies in your organization to [{setting_change}]."
```

variable interpolated values should be surrounded by [] for readability. :)

File: rules/auth0_rules/auth0_mfa_policy_enabled.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    )
+
+
+def alert_context(event):
+
```
Comment (andrea-youwakim) [inline rules/auth0_rules/auth0_mfa_policy_enabled.py:42]: please remove this unnecessary space

**Review by andrea-youwakim — CHANGES_REQUESTED** [review]
minor fmt change request, title change request to accommodate variable interpolation style requirements 

beautiful PR, thanks so much for this implementation and your participation in the hackathon, kleanthis!

File: rules/auth0_rules/auth0_mfa_policy_enabled.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+from global_filter_auth0 import filter_include_event
+from panther_auth0_helpers import auth0_alert_context, is_auth0_config_event
+from panther_base_helpers import deep_get
+
+
+def rule(event):
+
```
Comment (edyesed) [inline rules/auth0_rules/auth0_mfa_policy_enabled.py]:
```suggestion
def rule(event):
```

File: rules/auth0_rules/auth0_mfa_policy_enabled.yml [hunk 6]
```diff
@@ -0,0 +1,377 @@
+AnalysisType: rule
+Description: An Auth0 User enabled MFA Policy for your organization's tenant.
+DisplayName: "Auth0 MFA Policy Enabled"
+Enabled: true
+Filename: auth0_mfa_policy_enabled.py
+Runbook: Assess if this was done by the user for a valid business reason. Be vigilant to re-enable this setting as it's in the best security interest for your organization's security posture.
+Severity: Medium
+Tests:
+    - ExpectedResult: true
```
Comment (edyesed) [inline rules/auth0_rules/auth0_mfa_policy_enabled.yml:9]: will you please add a test case where `ExpectedResult: false`?

File: rules/auth0_rules/auth0_mfa_policy_enabled.yml [hunk 7]
```diff
@@ -0,0 +1,377 @@
+AnalysisType: rule
+Description: An Auth0 User enabled MFA Policy for your organization's tenant.
+DisplayName: "Auth0 MFA Policy Enabled"
+Enabled: true
+Filename: auth0_mfa_policy_enabled.py
+Runbook: Assess if this was done by the user for a valid business reason. Be vigilant to re-enable this setting as it's in the best security interest for your organization's security posture.
```
Comment (edyesed) [inline rules/auth0_rules/auth0_mfa_policy_enabled.yml:6]:
```suggestion
Runbook: Assess if this was done by the user for a valid business reason and was expected. This alert indicates a setting change that aligns with best security practices, follow-up may be unnecessary.
```

**Review by edyesed — CHANGES_REQUESTED** [review]
I recommend a runbook tweak, and adding a test case that indicates this detection will not alert on all events.

File: rules/auth0_rules/auth0_mfa_policy_enabled.py (hunk 1 above)
Comment (edyesed) [reply rules/auth0_rules/auth0_mfa_policy_enabled.py]: I pulled the branch and made this spacing change

File: rules/auth0_rules/auth0_mfa_policy_enabled.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (edyesed) [reply rules/auth0_rules/auth0_mfa_policy_enabled.py]: I pulled the branch and made this spacing change

File: rules/auth0_rules/auth0_mfa_policy_enabled.yml (hunk 6 above)
Comment (kleon919) [reply rules/auth0_rules/auth0_mfa_policy_enabled.yml:9]: Sure!

File: rules/auth0_rules/auth0_mfa_policy_enabled.yml (hunk 7 above)
Comment (kleon919) [reply rules/auth0_rules/auth0_mfa_policy_enabled.yml]: Done

File: rules/auth0_rules/auth0_mfa_policy_enabled.yml (hunk 6 above)
Comment (kleon919) [reply rules/auth0_rules/auth0_mfa_policy_enabled.yml:9]: Done

File: rules/auth0_rules/auth0_mfa_policy_enabled.py (hunk 4 above)
Comment (kleon919) [reply rules/auth0_rules/auth0_mfa_policy_enabled.py]: Done

**Comment by andrea-youwakim** [comment]
@kleon919 do you mind adding the test output in the pr description when you get the chance? This will help anyone viewing this pr to confirm desired detection functionality quickly and easily.

This can be done by running `pipenv run panther_analysis_tool test --filter RuleID=RuleID1` and replacing "RuleID1" with the rule id from this detection. 

Thanks so much!

**Comment by edyesed** [comment]
test output 


```text
user@computer:~/Sources/PAN/panther-analysis $ pipenv run panther_analysis_tool
test --filter RuleID=Auth0.MFA.Policy.Enabled
[INFO][root]: Testing analysis items in .

Auth0.MFA.Policy.Enabled
	[PASS] MFA Policy Enabled First
		[PASS] [rule] true
		[PASS] [title] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Always Require].
		[PASS] [dedup] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Always Require].
		[PASS] [alertContext] {"actor": {"email": "user.name@yourcompany.io", "name": "User Name", "user_id": "google-oauth2|105261262156475850461"}, "action": "Set the Multi-factor Authentication policies"}
	[PASS] MFA Policy Enabled Second
		[PASS] [rule] true
		[PASS] [title] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Use Adaptive MFA].
		[PASS] [dedup] Auth0 user [user.name@yourcompany.io] set the mfa policies in your organization to [Use Adaptive MFA].
		[PASS] [alertContext] {"actor": {"email": "user.name@yourcompany.io", "name": "User Name", "user_id": "google-oauth2|105261262156475850461"}, "action": "Set the Multi-factor Authentication policies"}
	[PASS] Other Event
		[PASS] [rule] false

--------------------------
Panther CLI Test Summary
	Path: .
	Passed: 1
	Failed: 0
	Invalid: 0

```

**Review by edyesed — APPROVED** [review]

**Review by edyesed — APPROVED** [review]
Thanks for the updates, good to ship

**Pull request merged**